temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowInternal.java

```diff
@@ -188,11 +189,24 @@ public static long currentTimeMillis() {
       Class<T> activityInterface,
       ActivityOptions options,
       Map<String, ActivityOptions> activityMethodOptions) {
+    // Merge the activity options we may have received from the workflow with the options we may
+    // have received in WorkflowImplementationOptions.
+    options = (options == null) ? getRootWorkflowContext().getDefaultActivityOptions() : options;
+    Map<String, ActivityOptions> mergedActivityOptionsMap = new HashMap<>();
+    if (getRootWorkflowContext().getActivityOptions() != null) {
+      mergedActivityOptionsMap.putAll(getRootWorkflowContext().getActivityOptions());
```

Comment:
getRootWorkflowContext is a relatively expensive operation that involves thread-local. I would call it once in this method. `WorkflowInternal.getWorkflowInterceptor()` down there can be also replaced with `context.getWorkflowInterceptor`.